### PR TITLE
[MM-38770] Show a copy button when hovering over code blocks

### PR DESCRIPTION
**Pull request by cleferman**
#### Summary

A copy button is now shown when hovering over a text block. Upon clicking it the code block gets copied into the clipboard and transitions into a checkmark to show copying the text was a success.

#### Ticket Link

Jira ticket: https://mattermost.atlassian.net/browse/MM-38770
Fixes https://github.com/mattermost/mattermost-server/issues/18851

#### Related Pull Requests

- Has server changes (please link here)
- Has mobile changes (please link here)

#### Release Note

```release-note
Added copy button when hovering over code blocks
```

### CODE REVIEW

**Review by hmhealey — APPROVED** [review]

**Comment by hmhealey** [comment]
Removing the Dev Review label since I forgot that we had a second review on the original PR

**Comment by matthewbirtch** [comment]
> copy button + label background on dark themes appears off. @matthewbirtch would you be able to provide a dark theme example of expected in figma? Thanks

@jgilliam17 this may actually be an issue with the `Indigo` theme. For some reason it's showing a white background for code blocks. Could you file a ticket to fix this?

This CSS seems to be setting the background and text colors for code
image

**Comment by jgilliam17** [comment]
@matthewbirtch  sounds good, I'll add a ticket for that.
Should we center the label vertically ? I feel it needs to be moved up few pixels 
Screen Shot 2022-01-26 at 2 43 42 PM

**Comment by matthewbirtch** [comment]
> Should we center the label vertically ? I feel it needs to be moved up few pixels

The background color behind the label and copy icon won't be distinguishable because the code background should be the same color - so you won't notice any off-centering. We could/should add bottom padding to that container though so it does sit vertically centered, but I think it was removed because it made it too high for single-line code

**Comment by jgilliam17** [comment]
Thanks Matt
-  Ticket added for Indigo theme codeblock background https://mattermost.atlassian.net/browse/MM-41299

**Comment by jgilliam17** [comment]
@cleferman 
One test is failing on Cypress, MM-T678 (see full report[ here](https://mattermost-cypress-report.s3.amazonaws.com/106352-57dcc47-onprem-ent-webapp-pr-9655-mattermost/mm-ee-test57dcc47/mochawesome.html))
posting this text, causes the white screen and console errors 
`/code 1. Not a list item, **not bolded**, http://notalink.com, ~off-topic is not a link to the channel.`
Screen Shot 2022-01-26 at 3 10 26 PM

**Comment by cleferman** [comment]
@jgilliam17 I'll have a look next week

**Comment by cleferman** [comment]
@jgilliam17 sorry it took so long. Please check :).
I'll be afk the entire week. If any other issues arise and it's needed urgently feel free to assign it to someone else until I get back. 
Cheers

**Comment by matthewbirtch** [comment]
@cleferman, I just had a quick look and it looks like the lines of code no longer align with the line numbers on the left
image

**Comment by cleferman** [comment]
@jgilliam17 I updated the HTML assertions for failing markdown tests. They were failing because the DOM changed to accommodate the copy button and the context menu. The `channel files search` test is unrelated to this feature. However, it seems to be passing right now (on my machine) without doing anything :).

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
@cleferman looks like the code lines and the numbers are aligning better now. Thanks!

What can we do to make the background of this container always the same color as the code area background? I can't seem to find the right variable for this.

153886711-4eb9fe58-208a-4d3b-88bf-cf75c92ec70c

**Comment by cleferman** [comment]
> @cleferman looks like the code lines and the numbers are aligning better now. Thanks!
> 
> What can we do to make the background of this container always the same color as the code area background? I can't seem to find the right variable for this.
> 
> 153886711-4eb9fe58-208a-4d3b-88bf-cf75c92ec70c

@matthewbirtch I can use --center-channel-big. That's what we use for the copy background.

I just checked all the themes with that color and the indigo theme has issues:
image

The text is black. I noticed that for Onyx it works, the the text is white. Should I push it like this? Not entirely sure what to do abut the Indigo theme, but seems like it's a theme issue.

**Comment by matthewbirtch** [comment]
> @matthewbirtch I can use --center-channel-big. That's what we use for the copy background.

I think the issue is that the code block background is using another color that is not coming from the main theme variables. I'm not sure how to leverage that color though

**Comment by matthewbirtch** [comment]
> I just checked all the themes with that color and the indigo theme has issues: image
> 
> The text is black. I noticed that for Onyx it works, the text is white. Should I push it like this? Not entirely sure what to do abut the Indigo theme, but seems like it's a theme issue.

Yes, I think another ticket was filed to resolve the indigo theme problem (https://mattermost.atlassian.net/browse/MM-41299)

**Comment by cleferman** [comment]
> > @matthewbirtch I can use --center-channel-big. That's what we use for the copy background.
> 
> I think the issue is that the code block background is using another color that is not coming from the main theme variables. I'm not sure how to leverage that color though
> 
> Screen Shot 2022-02-16 at 1 19 05 PM
> 
> h

I've put the background inside `.post-code .hljs` and it gets picked up because it's a stronger rule. Hope this is fine. Give it a go and let me know :).

**Comment by matthewbirtch** [comment]
> I've put the background inside `.post-code .hljs` and it gets picked up because it's a stronger rule. Hope this is fine. Give it a go and let me know :).

I think it's ok from my standpoint. However, it looks like there might be a typo `--center-channel-big` should be `--center-channel-bg`

In addition, can we adjust the padding on `post-code__overlay` to `padding: 4px 4px 0px;`? Right now the padding is causing it to overlap with the border when the code is only 1 line.

As mentioned, we'll need to get the issue with the Indigo theme fixed asap (ideally before this is merged).

**Comment by cleferman** [comment]
> > I've put the background inside `.post-code .hljs` and it gets picked up because it's a stronger rule. Hope this is fine. Give it a go and let me know :).
> 
> I think it's ok from my standpoint. However, it looks like there might be a typo `--center-channel-big` should be `--center-channel-bg`
> 
> In addition, can we adjust the padding on `post-code__overlay` to `padding: 4px 4px 0px;`? Right now the padding is causing it to overlap with the border when the code is only 1 line.
> 
> As mentioned, we'll need to get the issue with the Indigo theme fixed asap (ideally before this is merged).

Done.
I saw that the indigo issue is assigned to someone already.

File: sass/layout/_markdown.scss
Comment (matthewbirtch) [inline sass/layout/_markdown.scss:217]: change to `padding: 8.5px`

This will prevent the hover actions from spanning outside the bounds of single line code blocks

**Review by matthewbirtch — APPROVED** [review]
2 minor tweaks and we're good to go

**Comment by matthewbirtch** [comment]
> * @matthewbirtch do we need to add `do not merge` label while we wait for Indigo theme issue to be fixed?

@jgilliam17 I think so, yes. Otherwise we'll have a problem with code display using that theme.

**Comment by jgilliam17** [comment]
@cleferman 
[E2E report ](https://automation-dashboard.vercel.app/cycles/0424977e-8f36-4852-a753-838c37cb9f83)has failures related to code theming. This may be related to background change requested here. Can you please take a look? Thanks 🙂 
Screen Shot 2022-02-24 at 10 00 53 AM

**Comment by hmhealey** [comment]
We can wait for Matt to respond, but I think a revert of that definitely makes sense since it's breaking the code themes. Using a dark theme with the default GitHub code theme makes it basically unreadable

Screen Shot 2022-02-28 at 10 31 28 AM

**Comment by matthewbirtch** [comment]
> @jgilliam17 @matthewbirtch not sure what to do about these e2e tests. They basically expect a different background color for each theme color, but that background color is now --center-channel-bg.
> 
> So I guess it's the issue you were talking about earlier in this thread @matthewbirtch. image The colors coming from the theme should actually overwrite the main color variables because if they don't there's no guarantee we'll actually see the text with all these combinations we have. Should we revert [this](https://github.com/mattermost/mattermost-webapp/pull/9655#issuecomment-1041982327)?

Interesting, I'm not really sure how the code themes are incorporated. I guess we can't really override the code themes based on the comment from @hmhealey. Are we able to somehow pull the code theme's background color for the background of the `post-code__overlay` container?

**Comment by hmhealey** [comment]
I'm not actually sure about that. The way the code themes are applied is a bit of a black box. Unlike the regular theming where we have CSS variables, it's just a CSS file provided by the syntax highlighting library that we just download and apply to the whole page.

I suppose we could copy values from their CSS and add it as our own CSS variable. It would risk getting out of date, but that assumes highlight.js actually changes their style at any point

**Comment by esethna** [comment]
@cleferman let us know if you have a clear path forward on this?

**Comment by cleferman** [comment]
@esethna to be honest not really, I was waiting for a resolution on this. Meanwhile I reverted the change so the  background gets picked up from the theme.
Currently it looks like this (I guess it's similar to how it was before, the cypress tests linked here are passing):
image
I also added a fix for the copy container so it doesn't look like it's clipping when the code block is one liner:
image

**Comment by esethna** [comment]
Thanks @cleferman. @matthewbirtch @hmhealey I'm not super familiar with all the context here, can we suggest a path forward? ie the copy button BG could use the center channel BG color, and then we don't put the code type in that container (as in the original design, see below)? Apologies, I might not have all the context of past conversations

image

**Comment by matthewbirtch** [comment]
> Thanks @cleferman. @matthewbirtch @hmhealey I'm not super familiar with all the context here, can we suggest a path forward? ie when hovering, the copy button BG could use the center channel BG color, and then we don't need to put the code type in that container (as in the original design, see below)? Apologies, I might not have all the context of past conversations
> 
> image

@esethna that is how we originally had it (using the center channel bg color), but the code themes are the problem. The background used for the code container is different from the center channel background and seems to be dictated by the code theme provider (e.g. github, monokai, etc). The problem was especially highlighted with the current Indigo theme which uses the github code theme right now which is a light theme (but there is a ticket out to fix this). Even if we fix that, we run the problem when custom themes are selected using different code themes.

Based on what @hmhealey has said, I think what @cleferman has done with the most recent commits is all we can do.

> I suppose we could copy values from their CSS and add it as our own CSS variable. It would risk getting out of date, but that assumes highlight.js actually changes their style at any point

@hmhealey do you think this is one possible way to get the colors right?

**Comment by matthewbirtch** [comment]
> Thanks for the context. I'm not opposed to merging this as is (with maybe a few minor tweaks to padding etc via UX review), understanding that there is a UI flaw when there is a code theme that doesn't have the same BG color as the center channel. It's not the ideal outcome, but if we can't find a solution we're happy with then I wouldn't want to hold up the feature

Agreed! We could create a follow-up ticket to look in to improving this

**Comment by jgilliam17** [comment]
Thanks @cleferman 
Re-tested  - the only remaining issue is the background color of the container. I can create a follow up to address that after merge. @matthewbirtch Is this acceptable?
@hmhealey let me know if we want to try your suggestion to fix using CSS here in this PR or in a follow up ticket?
Thanks 🙂

**Comment by matthewbirtch** [comment]
> Re-tested - the only remaining issue is the background color of the container. I can create a follow up to address that after merge. @matthewbirtch Is this acceptable?

100% acceptable. Let's move forward with this and create a follow-up ticket to investigate

**Comment by hmhealey** [comment]
> @hmhealey do you think this is one possible way to get the colors right?

Yeah, that'd be one way of doing it, but it's not great since they could get out of sync when we update highlight.js. There might be a better way of doing that though

> @hmhealey let me know if we want to try your suggestion to fix using CSS here in this PR or in a follow up ticket?

Yeah, let's do that separately. My proposal is a bit hacky, so I'd like if we spent some more time seeing if there's a better way to do it

**Comment by jgilliam17** [comment]
I'll update and merge after github issues are resolved. 
Ticket added for follow up https://mattermost.atlassian.net/browse/MM-42594

**Review by jgilliam17 — APPROVED** [review]
Sorry about the delay I missed recent replies
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/172033-8b10702-onprem-ent-webapp-pr-9655-mm-ee-test:8b10702/mochawesome.html) also looks good, no PR related failures

**Comment by esethna** [comment]
@cleferman thanks for the PR! There was a bug found if you're keen on fixing it? https://mattermost.atlassian.net/browse/MM-42916

**Comment by cleferman** [comment]
@esethna I'll see if I can do something about it over the weekend. Is there a github issue already?

**Comment by amyblais** [comment]
There is a second bug ticket that @cleferman you may be interested to take: https://mattermost.atlassian.net/browse/MM-42777. cc @esethna

**Comment by esethna** [comment]
@cleferman thanks! We don't have github issues for these but feel free to just link the JIRA tickets in the PR description

**Comment by cleferman** [comment]
@amyblais I opened a PR for that here #10120.
@esethna Unfortunately I am not able to reproduce this. I was able to get my hands on a Macbook and tried reproducing it there as well on Safari 14.0.3, but no success. 
I noticed the context menu appears on a Mac if you hold ctrl while clicking on the trackpad, but I guess that's just the equivalent of a right click as I can't make a selection like that.

**Comment by amyblais** [comment]
> Unfortunately I am not able to reproduce this. I was able to get my hands on a Macbook and tried reproducing it there as well on Safari 14.0.3, but no success. I noticed the context menu appears on a Mac if you hold ctrl while clicking on the trackpad, but I guess that's just the equivalent of a right click as I can't make a selection like that.

@cleferman A dev reproduced it and added some comments to the ticket https://mattermost.atlassian.net/browse/MM-42916. Feel free to reach out to Harrison Healey if you're having trouble reproducing.

**Comment by cleferman** [comment]
@hmhealey the context menu feature is from the figma [mocks](https://www.figma.com/file/RV2tE26UUvoUlpx1mI4C32/MM-38770---Copy-Code-Blocks?node-id=1%3A47799). On the far left there's a screen called right-click menu.

I see it as well both on windows and on mac while long pressing without selecting anything. But not when selecting. From my perspective it's the right behaviour. (Maybe this is an issue on the newer version of Safari? I'm only able to test on 14.0.3) Check the gif below please. (The cursor turns yellow when I'm holding left-click)

context-menu-behaviour

I've used the `ContextMenuTrigger` which was used for a different feature. But now I see that I've set `holdToDisplay={1000}`  instead of `-1` like in the other place where this is used. Again, this does seem to work as intended, but if it's still happening on Safari I guess we can set it to -1.

**Comment by hmhealey** [comment]
Oh, I see it now. I didn't realize there was any new stuff on the left side of that design when I looked yesterday.

As for how it triggers, I think it's a distance thing. If you select text a few words and don't move your mouse very far, then it shows up, but if you move your mouse far enough before it would trigger, then it doesn't.

Either way though, I think we should just remove the `holdToDisplay` or set it to -1 would fix that for now unless we can find a way to make it only happen on a touch screen. I don't think it makes sense to include that behaviour when using a mouse or trackpad.

Let us know if you'd like to submit a PR for that. I think we're hoping to get that fixed in the next release, so we'd need something in in the next few days to get it for that. If not, no worries though.

**Comment by cleferman** [comment]
I'll submit a PR in a few

**Comment by esethna** [comment]
Thanks @cleferman !

**Pull request merged**
